### PR TITLE
feat(fgs/trigger):Add new fgs trigger support for function graph

### LINKED ISSUE #1215: Add support for creating a Trigger for FGS
I'd like to have the possibility to create a trigger for a FGS function.

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
Provide a resource to create a timing trigger with cron schedule or rate schedule.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1215

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add new trigger resource support.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

Timing trigger with rate schedule
```
make testacc TEST='./huaweicloud/services/acceptance/fgs' TESTARGS='-run=TestAccFunctionGraphTriggerV2_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/fgs -v -run=TestAccFunctionGraphTriggerV2_basic -timeout 360m -parallel 4
=== RUN   TestAccFunctionGraphTriggerV2_basic
=== PAUSE TestAccFunctionGraphTriggerV2_basic
=== CONT  TestAccFunctionGraphTriggerV2_basic
--- PASS: TestAccFunctionGraphTriggerV2_basic (20.26s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/fgs       20.420s
```
Timing trigger with cron schedule
```
make testacc TEST='./huaweicloud/services/acceptance/fgs' TESTARGS='-run=TestAccFunctionGraphTriggerV2_cronTimer'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/fgs -v -run=TestAccFunctionGraphTriggerV2_cronTimer -timeout 360m -parallel 4
=== RUN   TestAccFunctionGraphTriggerV2_cronTimer
=== PAUSE TestAccFunctionGraphTriggerV2_cronTimer
=== CONT  TestAccFunctionGraphTriggerV2_cronTimer
--- PASS: TestAccFunctionGraphTriggerV2_cronTimer (18.49s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/fgs       18.600s
```
OBS trigger
```
run=TestAccFunctionGraphTriggerV2_obs -timeout 360m -parallel 4
=== RUN   TestAccFunctionGraphTriggerV2_obs
=== PAUSE TestAccFunctionGraphTriggerV2_obs
=== CONT  TestAccFunctionGraphTriggerV2_obs
--- PASS: TestAccFunctionGraphTriggerV2_obs (44.29s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/fgs       44.353s
```
DIS Trigger
```
make testacc TEST='./huaweicloud/services/acceptance/fgs' TESTARGS='-run=TestAccFunctionGraphTriggerV2_dis'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/fgs -v -run=TestAccFunctionGraphTriggerV2_dis -timeout 360m -parallel 4
=== RUN   TestAccFunctionGraphTriggerV2_dis
=== PAUSE TestAccFunctionGraphTriggerV2_dis
=== CONT  TestAccFunctionGraphTriggerV2_dis
--- PASS: TestAccFunctionGraphTriggerV2_dis (64.12s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/fgs       64.181s
```
SMN trigger
```
make testacc TEST='./huaweicloud/services/acceptance/fgs' T
ESTARGS='-run=TestAccFunctionGraphTriggerV2_smn'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/fgs -v -run=TestAccFunctionGraphTriggerV2_smn -timeout 360m -parallel 4
=== RUN   TestAccFunctionGraphTriggerV2_smn
=== PAUSE TestAccFunctionGraphTriggerV2_smn
=== CONT  TestAccFunctionGraphTriggerV2_smn
--- PASS: TestAccFunctionGraphTriggerV2_smn (35.33s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/fgs        35.403s
```

### CODE REVIEW

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+				Required: true,
+				ForceNew: true,
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					timingTrigger,
+				}, false),
```
Comment (chengxiangdong) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go:70]: Valid values only support `TIMER`, we should set it as an optional parameter, and the set `Default: "TIMER"`

File: docs/resources/fgs_trigger.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+* `function_urn` - (Required, String, ForceNew) Specifies the Uniform Resource Name (URN) of the function.
+  Changing this will create a new trigger resource.
+
+* `type` - (Required, String, ForceNew) Specifies the type of the function.
+  The valid values currently only support __TIMER__.
+  Changing this will create a new trigger resource.
+
+* `status` - (Optional, String) Specifies whether trigger is enabled, default to true.
+
+* `timer` - (Optional, List, ForceNew) Specifies the configuration of the timing trigger.
```
Comment (chengxiangdong) [inline docs/resources/fgs_trigger.md:149]: optional or required ?

File: docs/resources/fgs_trigger.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  Changing this will create a new trigger resource.
+
+* `function_urn` - (Required, String, ForceNew) Specifies the Uniform Resource Name (URN) of the function.
+  Changing this will create a new trigger resource.
+
+* `type` - (Required, String, ForceNew) Specifies the type of the function.
+  The valid values currently only support __TIMER__.
+  Changing this will create a new trigger resource.
+
+* `status` - (Optional, String) Specifies whether trigger is enabled, default to true.
```
Comment (chengxiangdong) [inline docs/resources/fgs_trigger.md]: suggest to add valid values

File: docs/resources/fgs_trigger.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+* `timer` - (Optional, List, ForceNew) Specifies the configuration of the timing trigger.
+  Changing this will create a new trigger resource.
+  The `timer` object structure is documented below.
+
+The `timer` block supports:
```
Comment (chengxiangdong) [inline docs/resources/fgs_trigger.md:176]: missing the description of `additional_information`

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+func timerSchemaResource() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile("^([A-Za-z][A-Za-z0-9-_]{0,63})$"),
+					"The name can contains of 1 to 32 characters and start with a letter."+
```
Comment (chengxiangdong) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go]: s/"1 to 32"/"1 to 64"

File: docs/resources/fgs_trigger.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+* `name` - (Required, String, ForceNew) Specifies the trigger name, which can contains of 1 to 64 characters.
+  The name must start with a letter, only letters, digits, hyphens (-) and underscores (_) are allowed.
+  Changing this will create a new trigger resource.
+
+* `schedule_type` - (Required, String, ForceNew) Specifies the type of the time schedule.
+  The valid values are __Rate__ and __Corn__.
```
Comment (chengxiangdong) [inline docs/resources/fgs_trigger.md]: s/__Corn__/__Cron__

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+func resourceFunctionGraphTriggerV2Create(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*config.Config)
+	client, err := config.FgsV2Client(config.GetRegion(d))
+	if err != nil {
+		return fmtp.Errorf("Error creating HuaweiCloud FunctionGraph client: %s", err)
+	}
+
+	opts := buildFunctionGraphTriggerParameters(d, config)
+	log.Printf("[DEBUG] The create options is: %#v", opts)
+	urn := d.Get("function_urn").(string)
```
Comment (chengxiangdong) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go:379]: pls use 'logp'

File: docs/resources/fgs_trigger.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - resource ID in UUID format.
+
+## Import
+
+Triggers can be imported using their `id` and URN of the FunctionGraph to which the trigger belongs, separated by a slash, e.g.
+
+```
+$ terraform import huaweicloud_fgs_trigger.test 80e4640f-9e97-40ae-a787-8cb9b2f68528
```
Comment (chengxiangdong) [inline docs/resources/fgs_trigger.md]: paramter does not match "function_urn/id"

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go [hunk 9]
```diff
@@ -0,0 +1,260 @@
+package fgs
+
+import (
+	"log"
+	"regexp"
+	"strings"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
```
Comment (chengxiangdong) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go]: terraform-plugin-sdk has been upgraded to V2, pls pull the latest code from master

File: docs/resources/fgs_trigger.md (hunk 3 above)
Comment (Lance52259) [reply docs/resources/fgs_trigger.md]: The valid values has supported

File: docs/resources/fgs_trigger.md (hunk 4 above)
Comment (Lance52259) [reply docs/resources/fgs_trigger.md:176]: The missing parameter has been added

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go [hunk 10]
```diff
@@ -0,0 +1,624 @@
...
+
+	opts := trigger.CreateOpts{
+		TriggerTypeCode: triggerType,
+		TriggerStatus:   d.Get("status").(string),
+	}
+
+	switch triggerType {
+	case timingTrigger:
+		opts.EventData = buildTimingEventData(d)
+		opts.EventTypeCode = "MessageCreated"
```
Comment (ShiChangkuo) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go]: suggest to set **EventTypeCode** when initializing the CreateOpts.

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go [hunk 11]
```diff
@@ -0,0 +1,624 @@
...
+	}
+	logp.Printf("[DEBUG] The create options is: %#v", opts)
+	urn := d.Get("function_urn").(string)
+	resp, err := trigger.Create(client, opts, urn).Extract()
+	if err != nil {
+		return fmtp.Errorf("Error creating HuaweiCloud FunctionGraph trigger for function (%s): %s", urn, err)
+	}
+	d.SetId(resp.TriggerId)
+
+	if resp.TriggerTypeCode == kafkaTrigger && resp.Status != d.Get("status").(string) {
```
Comment (ShiChangkuo) [inline huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go]: suggest to add a default value for `status`.

File: huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_trigger_test.go [hunk 12]
```diff
@@ -0,0 +1,685 @@
...
+						"${huaweicloud_dms_kafka_instance.test.id}"),
+				),
+			},
+		},
+	})
+}
+
+func testAccFunctionGraphTimingTrigger_base(rName string) string {
+	return fmt.Sprintf(`
+resource "huaweicloud_fgs_function" "test" {
```
Comment (ShiChangkuo) [inline huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_trigger_test.go:296]: seems that we can extract huaweicloud_fgs_function as a base resource.

File: go.mod [hunk 13]
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/hashicorp/go-cleanhttp v0.5.2
 	github.com/hashicorp/go-multierror v1.0.0
 	github.com/hashicorp/terraform-plugin-sdk/v2 v2.7.0
+	github.com/huaweicloud/golangsdk v0.0.0-20210831081626-d823fe11ceba // indirect
```
Comment (ShiChangkuo) [inline go.mod]: please run `go mod tidy` command as we hace change the path of golangsdk package.

Comment (Lance52259) [reply go.mod]: already change repository to chnsz.

File: huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_trigger_test.go (hunk 12 above)
Comment (Lance52259) [reply huaweicloud/services/acceptance/fgs/resource_huaweicloud_fgs_trigger_test.go:296]: finished

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go (hunk 10 above)
Comment (Lance52259) [reply huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go]: The initialization `EventTypeCode` has been updated.

File: huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go (hunk 11 above)
Comment (Lance52259) [reply huaweicloud/services/fgs/resource_huaweicloud_fgs_trigger.go]: About empty setting of the status for DMS kafka, add some conditions to make trigger enter ACTIVE status successfully.

**Pull request merged**